lib/cc/engine/analyzers/file_thread_pool.rb

```diff
@@ -16,12 +16,13 @@ def run(&block)
           queue = build_queue
           lock = Mutex.new
 
-          @workers = thread_count.times.map do
-            Thread.new do
+          @workers = Array.new(thread_count) do
+            t = Thread.new do
               while (item = next_item(queue, lock))
                 yield item
               end
             end
+            t.abort_on_exception = true
```

Comment:
@brynary thanks! Experimented with `Thread#value` in a toy example and did find some useful information. I don't think it will work for us here.

My observation: when `#value` is called on a thread that contains a ruby exception, the exception is raised immediately. This is similar to the way we had expected a thread exception to get raised on `#join`.

In our case here though, the nature of the Java Exception issue is that it doesn't end up raising a ruby error. It just prints a stacktrace and continues. See https://github.com/jruby/jruby/issues/2354

So, the thread that encounters the Java level exception doesn't end up with the desired `#value`.

In my toy example, all of the threads end up with `nil` value and the final `"made past all the threads" line still gets printed:

```
require 'thread'
require 'securerandom'

threads = []
400.times do |n|
  threads << Thread.new do
    if n == 350
      leak = []
      loop do
        leak << SecureRandom.hex(3000)
      end
    else
      puts n
    end
  end
end

p threads.map(&:value).uniq

puts "made it past all the threads"
```

=>

```
...
398
399
Exception in thread "Ruby-0-Thread-353: app.rb:1" java.lang.OutOfMemoryError: Java heap space
	at org.jruby.util.ConvertBytes.twosComplementToUnsignedBytes(ConvertBytes.java:213)
	at org.jruby.util.ConvertBytes.twosComplementToHexBytes(ConvertBytes.java:176)
	at org.jruby.ext.securerandom.SecureRandomLibrary.hex(SecureRandomLibrary.java:31)
	at org.jruby.ext.securerandom.SecureRandomLibrary$INVOKER$s$hex.call(SecureRandomLibrary$INVOKER$s$hex.gen)
	at org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:161)
	at app.invokeOther1:hex(app.rb)
	at app.RUBY$block$\=app\,rb$2(app.rb:13)
	at java.lang.invoke.LambdaForm$DMH/1575540360.invokeStatic_L7_L(LambdaForm$DMH)
	at java.lang.invoke.LambdaForm$MH/1489069835.invokeExact_MT(LambdaForm$MH)
	at org.jruby.runtime.CompiledIRBlockBody.commonYieldPath(CompiledIRBlockBody.java:70)
	at org.jruby.runtime.IRBlockBody.yieldSpecific(IRBlockBody.java:67)
	at org.jruby.runtime.Block.yieldSpecific(Block.java:116)
	at org.jruby.RubyKernel.loop(RubyKernel.java:1291)
	at org.jruby.RubyKernel$INVOKER$s$0$0$loop.call(RubyKernel$INVOKER$s$0$0$loop.gen)
	at org.jruby.runtime.callsite.CachingCallSite.cacheAndCall(CachingCallSite.java:303)
	at org.jruby.runtime.callsite.CachingCallSite.callBlock(CachingCallSite.java:141)
	at org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:145)
	at app.invokeOther5:loop(app.rb)
	at app.RUBY$block$\=app\,rb$1(app.rb:12)
	at java.lang.invoke.LambdaForm$DMH/93805087.invokeStatic_L7_L(LambdaForm$DMH)
	at java.lang.invoke.LambdaForm$MH/1489069835.invokeExact_MT(LambdaForm$MH)
	at org.jruby.runtime.CompiledIRBlockBody.commonYieldPath(CompiledIRBlockBody.java:70)
	at org.jruby.runtime.IRBlockBody.call(IRBlockBody.java:59)
	at org.jruby.runtime.Block.call(Block.java:106)
	at org.jruby.RubyProc.call(RubyProc.java:334)
	at org.jruby.RubyProc.call(RubyProc.java:240)
	at org.jruby.internal.runtime.RubyRunnable.run(RubyRunnable.java:99)
	at java.lang.Thread.run(Thread.java:745)
[nil]
made it past all the threads
```